### PR TITLE
[PM-11894] update password generator policy

### PR DESCRIPTION
**Pull request by aj-rosado**
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-11894

## 📔 Objective

Renaming property `defaultValue` to `overridePasswordType` as it is the name used on the other clients
Updating the behaviour so that it forces the user to use the `overridePasswordType` on the password generator

## 📸 Screenshots

https://github.com/user-attachments/assets/4f021f4d-fd13-45fb-be9e-f709cb54a117

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: BitwardenShared/Core/Vault/Models/Enum/PolicyOptionType.swift [hunk 1]
```diff
@@ -10,7 +10,7 @@ enum PolicyOptionType: String {
     case capitalize
 
     /// A policy option for the default type of the password generator.
-    case defaultType
+    case overridePasswordType
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Vault/Models/Enum/PolicyOptionType.swift:13]: ⛏️ The documentation comment should probably be updated as well

File: BitwardenShared/Core/Vault/Services/PolicyService.swift [hunk 2]
```diff
@@ -179,15 +179,19 @@ extension DefaultPolicyService {
...
+               let overridePasswordType = PasswordGeneratorType(rawValue: overridePasswordTypeString),
+               generatorType != .password,
+               options.overridePasswordType != true {
                 // If there's multiple policies with different default types, the password type
                 // should take priority. Use `generateType` as opposed to `options.type` to ignore
                 // the existing type in the options.
-                generatorType = defaultType
+                generatorType = overridePasswordType
             }
+            // if the stringValue is NOT emptu or null, overridePasswordType should be true
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Vault/Services/PolicyService.swift:193]: ⛏️ 
```suggestion
            // if the stringValue is NOT empty or nil, overridePasswordType should be true
```

File: BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift [hunk 3]
```diff
@@ -48,6 +48,9 @@ struct PasswordGenerationOptions: Codable, Equatable {
 
     /// The separator to put between words in the passphrase.
     var wordSeparator: String?
+
+    /// Whether the password type should be enforced or not
+    var overridePasswordType: Bool?
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift:53]: 🤔 Why is this boolean optional? What does a nil value represent as different from a false value?

File: BitwardenShared/Core/Vault/Services/PolicyService.swift [hunk 4]
```diff
@@ -179,15 +179,19 @@ extension DefaultPolicyService {
...
+            options.overridePasswordType = !(policy[.overridePasswordType]?.stringValue?.isEmpty ?? true)
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Vault/Services/PolicyService.swift:194]: 🤔 Maybe I'm misunderstanding the logic, but doesn't putting this here mean that the `options.overridePasswordType != true` on line 187 will always be true for the first policy?

File: BitwardenShared/Core/Vault/Services/PolicyService.swift (hunk 2 above)
Comment (david-livefront) [reply BitwardenShared/Core/Vault/Services/PolicyService.swift:193]: Kotlin uses null not nil

File: BitwardenShared/Core/Vault/Models/Enum/PolicyOptionType.swift (hunk 1 above)
Comment (aj-rosado) [reply BitwardenShared/Core/Vault/Models/Enum/PolicyOptionType.swift:13]: updated!

File: BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift (hunk 3 above)
Comment (aj-rosado) [reply BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift]: should never be nil actually, removed the optional

File: BitwardenShared/Core/Vault/Services/PolicyService.swift (hunk 4 above)
Comment (aj-rosado) [reply BitwardenShared/Core/Vault/Services/PolicyService.swift]: you are right I messed up here.
It would work for most of the scenario but would fail on some edge cases. Fixed it and changed tests in order to catch this.

File: BitwardenShared/Core/Vault/Services/PolicyService.swift (hunk 2 above)
Comment (aj-rosado) [reply BitwardenShared/Core/Vault/Services/PolicyService.swift]: this comment was removed

File: BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift (hunk 3 above)
Comment (aj-rosado) [reply BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift]: Actually, this is making the [test_decode_empty](https://github.com/bitwarden/ios/blob/235f340f670d02f279a2409ec9bf596f63f888a3/BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptionsTests.swift#L54) on `PasswordGenerationOptionsTests` to fail.

I'm afraid it might throw an exception when calling the [decodePasswordOptions](https://github.com/bitwarden/ios/blob/235f340f670d02f279a2409ec9bf596f63f888a3/BitwardenShared/Core/Autofill/Utilities/ActionExtensionHelper.swift#L201) and there is no `overridePasswordType`.

File: BitwardenShared/UI/Tools/Generator/Generator/GeneratorProcessor.swift [hunk 5]
```diff
@@ -348,9 +348,7 @@ final class GeneratorProcessor: StateProcessor<GeneratorState, GeneratorAction,
         var passwordOptions = state.passwordState.passwordGenerationOptions
         state.isPolicyInEffect = await (try? services.policyService
             .applyPasswordGenerationPolicy(options: &passwordOptions)) ?? false
-        // When updating the state, don't update the password generator type or it can override the
-        // user's current selection.
-        state.passwordState.update(with: passwordOptions, shouldUpdateGeneratorType: false)
+        state.passwordState.update(with: passwordOptions, shouldUpdateGeneratorType: true)
```
Comment (fedemkr) [inline BitwardenShared/UI/Tools/Generator/Generator/GeneratorProcessor.swift:351]: 🤔 Can't the scenario in the comment happen anymore? I mean isn't this line with `true` value already called in `loadGeneratorOptions()`? Is it necessary to also have it here?

Comment (aj-rosado) [reply BitwardenShared/UI/Tools/Generator/Generator/GeneratorProcessor.swift:351]: This new behaviour should enforce the user to use the `generatorType` from policies, so the scenario we were trying to avoid, now it is mandatory.

This was causing sync to not update the `generatorType`

Comment (fedemkr) [reply BitwardenShared/UI/Tools/Generator/Generator/GeneratorProcessor.swift:351]: Great makes sense thanks!

**Review by KatherineInCode — COMMENTED** [review]
Seems reasonable so far, but we'll want to make a change so we don't have failing tests.

File: BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift (hunk 3 above)
Comment (matt-livefront) [reply BitwardenShared/Core/Tools/Models/Domain/PasswordGenerationOptions.swift]: This model is persisted to UserDefaults to maintain the user's options between app runs. The optional bool allows decoding to succeed even if it doesn't exist in the previously persisted model.

**Review by fedemkr — APPROVED** [review]
